// ipfs-cli/cjs/src/commands/files/chmod.js
'use strict';

var utils = require('../../utils.js');
var parseDuration = require('parse-duration');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var filesChmod = {
  command: 'chmod [mode] [path]',
  describe: 'Change file modes',
  builder: {
    path: {
      type: 'string',
      describe: 'The MFS path to change the mode of'
    },
    mode: {
      type: 'string',
      describe: 'The mode to use'
    },
    recursive: {
      alias: 'r',
      type: 'boolean',
      default: false,
      coerce: utils.asBoolean,
      describe: 'Whether to change modes recursively'
    },
    'hash-alg': {
      alias: 'h',
      type: 'string',
      default: 'sha2-256',
      describe: 'Hash function to use. Will set CID version to 1 if used'
    },
    flush: {
      alias: 'f',
      type: 'boolean',
      default: true,
      coerce: utils.asBoolean,
      describe: 'Flush the changes to disk immediately'
    },
    'shard-split-threshold': {
      type: 'number',
      default: 1000,
      describe: 'If a directory has more links than this, it will be transformed into a hamt-sharded-directory'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  handler({
    ctx: {ipfs},
    path,
    mode,
    recursive,
    hashAlg,
    flush,
    shardSplitThreshold,
    timeout
  }) {
    return ipfs.files.chmod(path, mode, {
      recursive,
      hashAlg,
      flush,
      shardSplitThreshold,
      timeout
    });
  }
};

module.exports = filesChmod;
