// ipfs-cli/cjs/src/commands/files/cp.js
'use strict';

var utils = require('../../utils.js');
var parseDuration = require('parse-duration');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var filesCp = {
  command: 'cp <source> <dest>',
  describe: 'Copy files between locations in the mfs',
  builder: {
    parents: {
      alias: 'p',
      type: 'boolean',
      default: false,
      coerce: utils.asBoolean,
      describe: 'Create any non-existent intermediate directories'
    },
    'hash-alg': {
      alias: 'h',
      type: 'string',
      default: 'sha2-256',
      describe: 'Hash function to use. Will set CID version to 1 if used'
    },
    flush: {
      alias: 'f',
      type: 'boolean',
      default: true,
      coerce: utils.asBoolean,
      describe: 'Flush the changes to disk immediately'
    },
    'shard-split-threshold': {
      type: 'number',
      default: 1000,
      describe: 'If a directory has more links than this, it will be transformed into a hamt-sharded-directory'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  handler({
    ctx: {ipfs},
    source,
    dest,
    parents,
    flush,
    hashAlg,
    shardSplitThreshold,
    timeout
  }) {
    return ipfs.files.cp(source, dest, {
      parents,
      flush,
      hashAlg,
      shardSplitThreshold,
      timeout
    });
  }
};

module.exports = filesCp;
