// ipfs-cli/cjs/src/commands/files/mkdir.js
'use strict';

var utils = require('../../utils.js');
var parseDuration = require('parse-duration');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var filesMkdir = {
  command: 'mkdir <path>',
  describe: 'Make mfs directories',
  builder: {
    parents: {
      alias: 'p',
      type: 'boolean',
      default: false,
      coerce: utils.asBoolean,
      describe: 'No error if existing, make parent directories as needed.'
    },
    'cid-version': {
      alias: ['cid-ver'],
      type: 'number',
      default: 0,
      describe: 'Cid version to use. (experimental).'
    },
    'hash-alg': {
      alias: 'h',
      type: 'string',
      default: 'sha2-256',
      describe: 'Hash function to use. Will set CID version to 1 if used'
    },
    flush: {
      alias: 'f',
      type: 'boolean',
      default: true,
      coerce: utils.asBoolean,
      describe: 'Flush the changes to disk immediately'
    },
    'shard-split-threshold': {
      type: 'number',
      default: 1000,
      describe: 'If a directory has more links than this, it will be transformed into a hamt-sharded-directory'
    },
    mode: {
      type: 'number',
      coerce: utils.asOctal,
      describe: 'Mode to apply to the new directory'
    },
    mtime: {
      type: 'number',
      coerce: utils.coerceMtime,
      describe: 'Modification time in seconds before or since the Unix Epoch to apply to created UnixFS entries'
    },
    'mtime-nsecs': {
      type: 'number',
      coerce: utils.coerceMtimeNsecs,
      describe: 'Modification time fraction in nanoseconds'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  handler({
    ctx: {ipfs},
    path,
    parents,
    cidVersion,
    hashAlg,
    flush,
    shardSplitThreshold,
    mode,
    mtime,
    mtimeNsecs,
    timeout
  }) {
    return ipfs.files.mkdir(path, {
      parents,
      cidVersion,
      hashAlg,
      flush,
      shardSplitThreshold,
      mode,
      mtime: utils.asMtimeFromSeconds(mtime, mtimeNsecs),
      timeout
    });
  }
};

module.exports = filesMkdir;
