// ipfs-cli/cjs/src/commands/files/touch.js
'use strict';

var utils = require('../../utils.js');
var parseDuration = require('parse-duration');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var filesTouch = {
  command: 'touch [path]',
  describe: 'change file modification times',
  builder: {
    mtime: {
      type: 'number',
      alias: 'm',
      coerce: utils.coerceMtime,
      describe: 'Modification time in seconds before or since the Unix Epoch to apply to created UnixFS entries'
    },
    'mtime-nsecs': {
      type: 'number',
      coerce: utils.coerceMtimeNsecs,
      describe: 'Modification time fraction in nanoseconds'
    },
    flush: {
      alias: 'f',
      type: 'boolean',
      default: true,
      coerce: utils.asBoolean,
      describe: 'Flush the changes to disk immediately'
    },
    'cid-version': {
      alias: ['cid-ver'],
      type: 'number',
      default: 0,
      describe: 'Cid version to use. (experimental).'
    },
    'hash-alg': {
      alias: 'h',
      type: 'string',
      default: 'sha2-256',
      describe: 'Hash function to use. Will set CID version to 1 if used'
    },
    'shard-split-threshold': {
      type: 'number',
      default: 1000,
      describe: 'If a directory has more links than this, it will be transformed into a hamt-sharded-directory'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  handler({
    ctx: {ipfs},
    path,
    flush,
    cidVersion,
    hashAlg,
    shardSplitThreshold,
    mtime,
    mtimeNsecs,
    timeout
  }) {
    return ipfs.files.touch(path, {
      mtime: utils.asMtimeFromSeconds(mtime, mtimeNsecs),
      flush,
      cidVersion,
      hashAlg,
      shardSplitThreshold,
      timeout
    });
  }
};

module.exports = filesTouch;
