// ipfs-cli/cjs/src/commands/object/patch/rm-link.js
'use strict';

var parseDuration = require('parse-duration');
var utils = require('../../../utils.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var objectPatchRmLink = {
  command: 'rm-link <root> <link>',
  describe: 'Remove a link from an object',
  builder: {
    root: {
      type: 'string',
      coerce: utils.coerceCID
    },
    'cid-base': {
      describe: 'Number base to display CIDs in. Note: specifying a CID base for v0 CIDs will have no effect.',
      type: 'string',
      default: 'base58btc'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({
    ctx: {ipfs, print},
    root,
    link,
    cidBase,
    timeout
  }) {
    const cid = await ipfs.object.patch.rmLink(root, link, { timeout });
    const base = await ipfs.bases.getBase(cidBase);
    print(cid.toString(base.encoder));
  }
};

module.exports = objectPatchRmLink;
