// ipfs-cli/cjs/src/commands/repo/gc.js
'use strict';

var parseDuration = require('parse-duration');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var repoGc = {
  command: 'gc',
  describe: 'Perform a garbage collection sweep on the repo.',
  builder: {
    quiet: {
      alias: 'q',
      desc: 'Write minimal output',
      type: 'boolean',
      default: false
    },
    'stream-errors': {
      desc: 'Output individual errors thrown when deleting blocks.',
      type: 'boolean',
      default: true
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({
    ctx: {ipfs, print},
    quiet,
    streamErrors,
    timeout
  }) {
    for await (const r of ipfs.repo.gc({ timeout })) {
      if (r.err != null) {
        streamErrors && print(r.err.message, true, true);
      } else {
        print((quiet ? '' : 'removed ') + r.cid);
      }
    }
  }
};

module.exports = repoGc;
