// ipfs-cli/cjs/src/commands/stats/bw.js
'use strict';

var parseDuration = require('parse-duration');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var statsBw = {
  command: 'bw',
  describe: 'Get bandwidth information.',
  builder: {
    peer: {
      type: 'string',
      default: ''
    },
    proto: {
      type: 'string',
      default: ''
    },
    poll: {
      type: 'boolean',
      default: false
    },
    interval: {
      type: 'string',
      default: '1s',
      coerce: parseDuration__default["default"]
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({
    ctx: {ipfs, print},
    peer,
    proto,
    poll,
    interval,
    timeout
  }) {
    for await (const chunk of ipfs.stats.bw({
        peer,
        proto,
        poll,
        interval,
        timeout
      })) {
      print(`bandwidth status
  total in: ${ chunk.totalIn }B
  total out: ${ chunk.totalOut }B
  rate in: ${ chunk.rateIn }B/s
  rate out: ${ chunk.rateOut }B/s`);
    }
  }
};

module.exports = statsBw;
