// ipfs-http-server/cjs/src/api/resources/resolve.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var joi = require('../../utils/joi.js');

const resolveResource = {
  options: {
    validate: {
      options: {
        allowUnknown: true,
        stripUnknown: true
      },
      query: joi.object().keys({
        path: joi.string().required(),
        recursive: joi.boolean().default(true),
        cidBase: joi.string().default('base58btc'),
        timeout: joi.timeout()
      }).rename('arg', 'path', {
        override: true,
        ignoreUndefined: true
      }).rename('cid-base', 'cidBase', {
        override: true,
        ignoreUndefined: true
      })
    }
  },
  async handler(request, h) {
    const {
      app: {signal},
      server: {
        app: {ipfs}
      },
      query: {path, recursive, cidBase, timeout}
    } = request;
    const res = await ipfs.resolve(path, {
      recursive,
      cidBase,
      signal,
      timeout
    });
    return h.response({ Path: res });
  }
};

exports.resolveResource = resolveResource;
