// ipfs-grpc-server/cjs/src/endpoints/add.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var pushable = require('it-pushable');
var itPipe = require('it-pipe');
var encodeMtime = require('../utils/encode-mtime.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var pushable__default = /*#__PURE__*/_interopDefaultLegacy(pushable);

function grpcAdd(ipfs, options = {}) {
  async function add(source, sink, metadata) {
    const opts = {
      ...metadata,
      progress: (bytes = 0, path = '') => {
        sink.push({
          type: 'PROGRESS',
          bytes,
          path
        });
      }
    };
    await itPipe.pipe(async function* toInput() {
      const fileInputStream = pushable__default["default"]();
      setTimeout(async () => {
        const streams = [];
        try {
          for await (const {index, type, path, mode, mtime, mtimeNsecs, content} of source) {
            let mtimeObj;
            if (mtime != null) {
              mtimeObj = {
                secs: mtime,
                nsecs: undefined
              };
              if (mtimeNsecs != null) {
                mtimeObj.nsecs = mtimeNsecs;
              }
            }
            if (!type || type === 'DIRECTORY') {
              fileInputStream.push({
                path,
                mode: mode !== 0 ? mode : undefined,
                mtime: mtimeObj
              });
              continue;
            }
            let stream = streams[index];
            if (!stream) {
              stream = streams[index] = pushable__default["default"]();
              fileInputStream.push({
                path,
                mode: mode !== 0 ? mode : undefined,
                mtime: mtimeObj,
                content: stream
              });
            }
            if (content && content.length) {
              stream.push(content);
            } else {
              stream.end();
              streams[index] = null;
            }
          }
          fileInputStream.end();
        } catch (err) {
          fileInputStream.end(err);
        } finally {
          streams.forEach(stream => stream && stream.end());
        }
      }, 0);
      yield* fileInputStream;
    }, async function (source) {
      for await (const result of ipfs.addAll(source, opts)) {
        sink.push({
          ...result,
          type: 'RESULT',
          cid: result.cid.toString(),
          ...encodeMtime.encodeMtime(result.mtime)
        });
      }
      sink.end();
    });
  }
  return add;
}

exports.grpcAdd = grpcAdd;
