// ipfs-cli/cjs/src/commands/object/links.js
'use strict';

var parseDuration = require('parse-duration');
var utils = require('../../utils.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var objectLinks = {
  command: 'links <key>',
  describe: 'Outputs the links pointed to by the specified object',
  builder: {
    key: {
      type: 'string',
      coerce: utils.coerceCID
    },
    'cid-base': {
      describe: 'Number base to display CIDs in. Note: specifying a CID base for v0 CIDs will have no effect.',
      type: 'string',
      default: 'base58btc'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({
    ctx: {ipfs, print},
    key,
    cidBase,
    timeout
  }) {
    const links = await ipfs.object.links(key, { timeout });
    const base = await ipfs.bases.getBase(cidBase);
    links.forEach(link => {
      const cidStr = link.Hash.toString(base.encoder);
      print(`${ cidStr } ${ link.Tsize } ${ utils.stripControlCharacters(link.Name) }`);
    });
  }
};

module.exports = objectLinks;
