// ipfs-cli/cjs/src/commands/bootstrap/add.js
'use strict';

var parseDuration = require('parse-duration');
var utils = require('../../utils.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var bootstrapAdd = {
  command: 'add [<peer>]',
  describe: 'Add peers to the bootstrap list',
  builder: {
    peer: {
      type: 'string',
      coerce: utils.coerceMultiaddr
    },
    default: {
      describe: 'Add default bootstrap nodes.',
      type: 'boolean',
      default: false
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({
    ctx: {ipfs, print},
    peer,
    default: defaultPeers,
    timeout
  }) {
    let list;
    if (peer) {
      list = await ipfs.bootstrap.add(peer, { timeout });
    } else if (defaultPeers) {
      list = await ipfs.bootstrap.reset({ timeout });
    } else {
      throw new Error('Please specify a peer or the --default flag');
    }
    list.Peers.forEach(peer => print(peer.toString()));
  }
};

module.exports = bootstrapAdd;
