// ipfs-cli/cjs/src/commands/bootstrap/rm.js
'use strict';

var parseDuration = require('parse-duration');
var utils = require('../../utils.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var bootstrapRm = {
  command: 'rm [<peer>]',
  describe: 'Removes peers from the bootstrap list',
  builder: {
    peer: {
      type: 'string',
      coerce: utils.coerceMultiaddr
    },
    all: {
      type: 'boolean',
      describe: 'Remove all bootstrap peers.',
      default: false
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({
    ctx: {ipfs, print},
    all,
    peer,
    timeout
  }) {
    let list;
    if (peer) {
      list = await ipfs.bootstrap.rm(peer, { timeout });
    } else if (all) {
      list = await ipfs.bootstrap.clear({ timeout });
    } else {
      throw new Error('Please specify a peer or the --all flag');
    }
    list.Peers.forEach(peer => print(peer.toString()));
  }
};

module.exports = bootstrapRm;
