// ipfs-grpc-server/cjs/src/utils/load-services.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var protocol = require('ipfs-grpc-protocol');
var protobuf = require('protobufjs/light.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var protocol__default = /*#__PURE__*/_interopDefaultLegacy(protocol);
var protobuf__default = /*#__PURE__*/_interopDefaultLegacy(protobuf);

const {Service} = protobuf__default["default"];
const CONVERSION_OPTS = {
  keepCase: false,
  longs: String,
  enums: String,
  defaults: false,
  oneofs: true
};
function loadServices() {
  const root = protobuf__default["default"].Root.fromJSON(protocol__default["default"]);
  const output = {};
  Object.keys(root.nested.ipfs).filter(key => root.nested.ipfs[key] instanceof Service).map(key => root.nested.ipfs[key]).forEach(service => {
    const serviceDef = {};
    output[service.name] = serviceDef;
    Object.keys(service.methods).forEach(methodName => {
      const method = service.methods[methodName].resolve();
      serviceDef[methodName] = {
        path: `/ipfs.${ service.name }/${ methodName }`,
        requestStream: method.requestStream,
        responseStream: method.responseStream,
        requestSerialize: obj => {
          const message = method.resolvedRequestType.fromObject(obj);
          return method.resolvedRequestType.encode(message).finish();
        },
        requestDeserialize: buf => {
          const message = method.resolvedRequestType.decode(buf);
          const obj = method.resolvedRequestType.toObject(message, CONVERSION_OPTS);
          Object.defineProperty(obj, 'toObject', {
            enumerable: false,
            configurable: false,
            value: () => obj
          });
          return obj;
        },
        responseSerialize: obj => {
          const message = method.resolvedResponseType.fromObject(obj);
          return method.resolvedResponseType.encode(message).finish();
        },
        responseDeserialize: buf => {
          const message = method.resolvedResponseType.decode(buf);
          const obj = method.resolvedResponseType.toObject(message, CONVERSION_OPTS);
          Object.defineProperty(obj, 'toObject', {
            enumerable: false,
            configurable: false,
            value: () => obj
          });
          return obj;
        }
      };
    });
  });
  return output;
}

exports.loadServices = loadServices;
