// ipfs-cli/cjs/src/commands/cat.js
'use strict';

var parseDuration = require('parse-duration');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var cat = {
  command: 'cat <ipfsPath>',
  describe: 'Fetch and cat an IPFS path referencing a file',
  builder: {
    offset: {
      alias: 'o',
      type: 'integer',
      describe: 'Byte offset to begin reading from'
    },
    length: {
      alias: [
        'n',
        'count'
      ],
      type: 'integer',
      describe: 'Maximum number of bytes to read'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({
    ctx: {ipfs, print},
    ipfsPath,
    offset,
    length,
    timeout
  }) {
    for await (const buf of ipfs.cat(ipfsPath, {
        offset,
        length,
        timeout
      })) {
      print.write(buf);
    }
  }
};

module.exports = cat;
