// ipfs-cli/cjs/src/commands/bitswap/unwant.js
'use strict';

var parseDuration = require('parse-duration');
var utils = require('../../utils.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var bitswapUnwant = {
  command: 'unwant <key>',
  describe: 'Removes a given block from your wantlist.',
  builder: {
    key: {
      alias: 'k',
      describe: 'Key to remove from your wantlist',
      type: 'string',
      coerce: utils.coerceCID
    },
    'cid-base': {
      describe: 'Number base to display CIDs in. Note: specifying a CID base for v0 CIDs will have no effect.',
      type: 'string',
      default: 'base58btc'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({ctx, key, cidBase, timeout}) {
    const {ipfs, print} = ctx;
    const base = await ipfs.bases.getBase(cidBase);
    await ipfs.bitswap.unwant(key, { timeout });
    print(`Key ${ key.toString(base.encoder) } removed from wantlist`);
  }
};

module.exports = bitswapUnwant;
