// ipfs-cli/cjs/src/commands/bitswap/wantlist.js
'use strict';

var parseDuration = require('parse-duration');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var bitswapWantlist = {
  command: 'wantlist [peer]',
  describe: 'Print out all blocks currently on the bitswap wantlist for the local peer.',
  builder: {
    peer: {
      alias: 'p',
      describe: 'Specify which peer to show wantlist for.',
      type: 'string'
    },
    'cid-base': {
      describe: 'Number base to display CIDs in. Note: specifying a CID base for v0 CIDs will have no effect.',
      type: 'string',
      default: 'base58btc'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({ctx, peer, cidBase, timeout}) {
    const {ipfs, print} = ctx;
    const base = await ipfs.bases.getBase(cidBase);
    let list;
    if (peer) {
      list = await ipfs.bitswap.wantlistForPeer(peer, { timeout });
    } else {
      list = await ipfs.bitswap.wantlist({ timeout });
    }
    list.forEach(cid => print(cid.toString(base.encoder)));
  }
};

module.exports = bitswapWantlist;
