// ipfs-cli/cjs/src/commands/key/export.js
'use strict';

var fs = require('fs');
var parseDuration = require('parse-duration');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);
var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var keyExport = {
  command: 'export <name>',
  describe: 'Export the key as a password protected PKCS #8 PEM file',
  builder: {
    passout: {
      alias: 'p',
      describe: 'Password for the PEM',
      type: 'string',
      demandOption: true
    },
    output: {
      alias: 'o',
      describe: 'Output file',
      type: 'string',
      default: 'stdout'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({ctx, name, passout, output, timeout}) {
    const {ipfs} = ctx;
    const pem = await ipfs.key.export(name, passout, { timeout });
    if (output === 'stdout') {
      process.stdout.write(pem);
    } else {
      fs__default["default"].writeFileSync(output, pem);
    }
  }
};

module.exports = keyExport;
