// ipfs-cli/cjs/src/commands/key/gen.js
'use strict';

var parseDuration = require('parse-duration');
var utils = require('../../utils.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var keyGen = {
  command: 'gen <name>',
  describe: 'Create a new key',
  builder: {
    type: {
      alias: 't',
      describe: 'type of the key to create',
      choices: [
        'rsa',
        'ed25519'
      ],
      default: 'ed25519'
    },
    size: {
      alias: 's',
      describe: 'size of the key to generate.',
      default: 2048,
      type: 'number'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({
    ctx: {ipfs, print},
    name,
    type,
    size,
    timeout
  }) {
    const key = await ipfs.key.gen(name, {
      type,
      size,
      timeout
    });
    print(`generated ${ key.id } ${ utils.stripControlCharacters(key.name) }`);
  }
};

module.exports = keyGen;
