// ipfs-cli/cjs/src/commands/key/import.js
'use strict';

var fs = require('fs');
var parseDuration = require('parse-duration');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);
var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var keyImport = {
  command: 'import <name>',
  describe: 'Import the key from a PKCS #8 PEM file',
  builder: {
    passin: {
      alias: 'p',
      describe: 'Password for the PEM',
      type: 'string'
    },
    input: {
      alias: 'i',
      describe: 'Input PEM file',
      type: 'string',
      demandOption: true,
      coerce: input => fs__default["default"].readFileSync(input, 'utf8')
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({
    ctx: {ipfs, print},
    name,
    input,
    passin,
    timeout
  }) {
    const key = await ipfs.key.import(name, input, passin, { timeout });
    print(`imported ${ key.id } ${ key.name }`);
  }
};

module.exports = keyImport;
