// ipfs-cli/cjs/src/commands/block/stat.js
'use strict';

var parseDuration = require('parse-duration');
var utils = require('../../utils.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var blockStat = {
  command: 'stat <key>',
  describe: 'Print information of a raw IPFS block',
  builder: {
    key: {
      type: 'string',
      coerce: utils.coerceCID
    },
    'cid-base': {
      describe: 'Number base to display CIDs in.',
      type: 'string',
      default: 'base58btc'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({ctx, key, cidBase, timeout}) {
    const {ipfs, print} = ctx;
    const stats = await ipfs.block.stat(key, { timeout });
    const base = await ipfs.bases.getBase(cidBase);
    print('Key: ' + stats.cid.toString(base.encoder));
    print('Size: ' + stats.size);
  }
};

module.exports = blockStat;
