// ipfs-http-server/cjs/src/api/resources/files/mv.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var joi = require('../../../utils/joi.js');

const mvResource = {
  options: {
    validate: {
      options: {
        allowUnknown: true,
        stripUnknown: true
      },
      query: joi.object().keys({
        arg: joi.array().required().items(joi.string()).min(2),
        recursive: joi.boolean().default(false),
        parents: joi.boolean().default(false),
        hashAlg: joi.string().default('sha2-256'),
        cidVersion: joi.number().integer().valid(0, 1).default(0),
        flush: joi.boolean().default(true),
        shardSplitThreshold: joi.number().integer().min(0).default(1000),
        timeout: joi.timeout()
      }).rename('shard-split-threshold', 'shardSplitThreshold', {
        override: true,
        ignoreUndefined: true
      }).rename('hash-alg', 'hashAlg', {
        override: true,
        ignoreUndefined: true
      }).rename('hash', 'hashAlg', {
        override: true,
        ignoreUndefined: true
      }).rename('cid-version', 'cidVersion', {
        override: true,
        ignoreUndefined: true
      })
    }
  },
  async handler(request, h) {
    const {ipfs} = request.server.app;
    const {arg, recursive, parents, hashAlg, cidVersion, flush, shardSplitThreshold, timeout} = request.query;
    const args = arg.concat({
      recursive,
      parents,
      cidVersion,
      flush,
      hashAlg,
      shardSplitThreshold,
      signal: request.app.signal,
      timeout
    });
    await ipfs.files.mv.apply(null, args);
    return h.response();
  }
};

exports.mvResource = mvResource;
