// ipfs-http-server/cjs/src/api/resources/files/read.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var joi = require('../../../utils/joi.js');
var streamResponse = require('../../../utils/stream-response.js');

const readResource = {
  options: {
    validate: {
      options: {
        allowUnknown: true,
        stripUnknown: true
      },
      query: joi.object().keys({
        arg: joi.string().required(),
        offset: joi.number().integer().min(0),
        length: joi.number().integer().min(0),
        timeout: joi.timeout()
      }).rename('o', 'offset', {
        override: true,
        ignoreUndefined: true
      }).rename('n', 'length', {
        override: true,
        ignoreUndefined: true
      }).rename('count', 'length', {
        override: true,
        ignoreUndefined: true
      })
    }
  },
  handler(request, h) {
    const {ipfs} = request.server.app;
    const {arg, offset, length, timeout} = request.query;
    return streamResponse.streamResponse(request, h, () => ipfs.files.read(arg, {
      offset,
      length,
      signal: request.app.signal,
      timeout
    }));
  }
};

exports.readResource = readResource;
