// ipfs-http-server/cjs/src/api/resources/files/rm.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var joi = require('../../../utils/joi.js');

const rmResource = {
  options: {
    validate: {
      options: {
        allowUnknown: true,
        stripUnknown: true
      },
      query: joi.object().keys({
        arg: joi.array().required().items(joi.string()).min(1).single(),
        recursive: joi.boolean().default(false),
        shardSplitThreshold: joi.number().integer().min(0).default(1000),
        timeout: joi.timeout()
      }).rename('r', 'recursive', {
        override: true,
        ignoreUndefined: true
      }).rename('shard-split-threshold', 'shardSplitThreshold', {
        override: true,
        ignoreUndefined: true
      })
    }
  },
  async handler(request, h) {
    const {
      server: {
        app: {ipfs}
      },
      query: {arg, recursive, shardSplitThreshold, timeout},
      app: {signal}
    } = request;
    await ipfs.files.rm(arg, {
      recursive,
      shardSplitThreshold,
      signal,
      timeout
    });
    return h.response();
  }
};

exports.rmResource = rmResource;
