// ipfs-http-server/cjs/src/api/resources/files/touch.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var joi = require('../../../utils/joi.js');
var parseMtime = require('./utils/parse-mtime.js');

const touchResource = {
  options: {
    validate: {
      options: {
        allowUnknown: true,
        stripUnknown: true
      },
      query: joi.object().keys({
        arg: joi.string().required(),
        mtime: joi.number().integer(),
        mtimeNsecs: joi.number().integer().min(0),
        hashAlg: joi.string().default('sha2-256'),
        cidVersion: joi.number().integer().valid(0, 1).default(0),
        flush: joi.boolean().default(true),
        shardSplitThreshold: joi.number().integer().min(0).default(1000),
        timeout: joi.timeout()
      }).rename('shard-split-threshold', 'shardSplitThreshold', {
        override: true,
        ignoreUndefined: true
      }).rename('mtime-nsecs', 'mtimeNsecs', {
        override: true,
        ignoreUndefined: true
      }).rename('hash-alg', 'hashAlg', {
        override: true,
        ignoreUndefined: true
      }).rename('hash', 'hashAlg', {
        override: true,
        ignoreUndefined: true
      }).rename('cid-version', 'cidVersion', {
        override: true,
        ignoreUndefined: true
      })
    }
  },
  async handler(request, h) {
    const {ipfs} = request.server.app;
    const {arg, flush, shardSplitThreshold, cidVersion, hashAlg, mtime, mtimeNsecs, timeout} = request.query;
    await ipfs.files.touch(arg, {
      mtime: parseMtime.parseMtime(mtime, mtimeNsecs),
      flush,
      shardSplitThreshold,
      cidVersion,
      hashAlg,
      signal: request.app.signal,
      timeout
    });
    return h.response();
  }
};

exports.touchResource = touchResource;
