// ipfs-cli/cjs/src/commands/dht/find-providers.js
'use strict';

var parseDuration = require('parse-duration');
var utils = require('../../utils.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var parseDuration__default = /*#__PURE__*/_interopDefaultLegacy(parseDuration);

var dhtFindProviders = {
  command: 'findprovs <key>',
  describe: 'Find peers that can provide a specific value, given a key.',
  builder: {
    key: {
      type: 'string',
      coerce: utils.coerceCID
    },
    'num-providers': {
      alias: 'n',
      describe: 'The number of providers to find. Default: 20.',
      default: 20,
      type: 'number'
    },
    timeout: {
      type: 'string',
      coerce: parseDuration__default["default"]
    }
  },
  async handler({
    ctx: {ipfs, print},
    key,
    numProviders,
    timeout
  }) {
    const providers = new Set();
    for await (const event of ipfs.dht.findProvs(key, { timeout })) {
      if (event.name === 'PROVIDER') {
        event.providers.forEach(peerData => {
          if (providers.has(peerData.id)) {
            return;
          }
          providers.add(peerData.id);
          print(peerData.id.toString());
        });
        if (providers.size >= numProviders) {
          break;
        }
      }
    }
  }
};

module.exports = dhtFindProviders;
